### PR TITLE
Demonstrate back-pressure in decoupled interface testing

### PR DESCRIPTION
**Pull request by chick**
Adds a test that illustrates back-pressure input stalling by occasionally inserting
extra steps in enqueue and dequeue blocks

Based on [BootCamp PR 153](https://github.com/freechipsproject/chisel-bootcamp/pull/153)

### CODE REVIEW

**Comment by chick** [comment]
This replaces #306 with a cleaner interface and a
test with decent examples.

**Comment by ducky64** [comment]
Are you sure these are necessary? The current implementation of [`initSource` ](https://github.com/ucb-bar/chisel-testers2/blob/master/src/main/scala/chiseltest/DecoupledDriver.scala#L12) and [`initSink`](https://github.com/ucb-bar/chisel-testers2/blob/master/src/main/scala/chiseltest/DecoupledDriver.scala#L62) set valid and ready to false, which is intended to be the default state unless temporarily overridden by something like `enqueue` or `expectDequeue`.

Also, are you sure about the implementation? The idea of a timescope is that the pokes inside disappear (and revert) at the end of a timescope. As written, it seems you open a timescope, do a poke, and immediately close the timescope - so it effectively shouldn't do anything. Typically the idea with a timescope is that it should do some kind of operation that advances time, like stepping on a block.

Note that this PR seems to follows the common latching pokes style of writing tests, where you would have to remember to poke (for example) valid = false after asserting valid = true for one transaction. In contrast, the chiseltest style provides a default and a time-scoped temporary override, with the system handling the revert for you.

**Comment by chick** [comment]
@ducky64 Several points:
- I could not get this to work without wrapping using `timescope`s in `notReady` and `notValid`
  - I was unable to get the tests in `DecoupledBackPressureInputStallTest` to work using `initSource` and `initSink`
  - Maybe you could provide an example of how you see the API on this.
- Motivation for this was @wunderabt trying to construct a decoupled back-pressure example for the bootcamp
  - They tried using a slower clock in the module to create the back-pressure but that approach has a lot of problems.
- I'm a big believer in TBUINU "the best UI is no UI", it follows
  - Back-pressure should be built-in to the decoupled tools in some way that is clear and painless
  - Maybe `enqueue`, `dequeue` family could have additional arguments, with good defaults.
  - Maybe @ekiwi has some ideas here.
    - We should make it hard to write tests of decoupled interfaces that don't handle back-pressure
- If this turns out to be really complicated I think we should make some small step that will let us get a bootcamp example

**Comment by chick** [comment]
These errors should be fixed by PR #329

**Review by ekiwi — APPROVED** [review]
LGTM

File: src/test/scala/chiseltest/tests/DecoupledBackPressureInputStallTest.scala [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+      x := bundle.value1
+      y := bundle.value2
+      xInitial := bundle.value1
+      yInitial := bundle.value2
+      busy := true.B
+      output.valid := false.B
+    }
+  }
+
+  val cycles = IO(Output(UInt(32.W)))
```
Comment (ducky64) [inline src/test/scala/chiseltest/tests/DecoupledBackPressureInputStallTest.scala:75]: In a proper design, this wouldn't be part of the DUT right? This is pure test instrumentation? Can this be moved into the test harness, or if a cycle count isn't available yet (is it?) can we have a TODO for it?

**Review by ducky64 — CHANGES_REQUESTED** [review]
Mostly looks good, some stylistic issues.

**Comment by chick** [comment]
Closing as this is more of an example than a test. This was prompted by [Bootcamp PR 153](https://github.com/freechipsproject/chisel-bootcamp/pull/153). 153 has been closed as I believe the DecoupledGCD code demonstrates back-pressure handling already

**Pull request closed without merging**